src/XMakeBuildEngine/Evaluation/Evaluator.cs

```diff
@@ -2026,15 +2021,164 @@ private bool EvaluateWhenOrOtherwiseChildren(IEnumerable<ProjectElement> childre
+
+            // $(MSBuildExtensionsPath*) usually resolves to a single value, single default path. This can be overridden
+            // by the usual property overriding techniques.
+            //
+            // Eg. <Import Project='$(MSBuildExtensionsPath)\foo\extn.proj' />
+            //
+            // But this feature allows that when it is used in an Import element, it will behave as a "search path", meaning
+            // that the relative project path "foo\extn.proj" will be searched for, in more than one location.
+            // Essentially, we will try to load that project file by trying multiple values (search paths) for the
+            // $(MSBuildExtensionsPath*) property.
+            //
+            // The various paths tried, in order are:
+            //
+            // 1. The value of the MSBuildExtensionsPath* property
+            //
+            // 2. /Library/Frameworks/Mono.framework/External/xbuild (only on OSX)
+            //
+            // 3. One or more paths specified via the MSBuildExtensionsPathSearchPaths *environment variable* (this works for
+            //    MSBuildExtensionsPath32SearchPaths and MSBuildExtensionsPath64SearchPaths properties too, for their corresponding
+            //    msbuild properties.
+            //
+            // This is available only when used in an Import element and it's Condition. So, the following common pattern
+            // would work:
+            //
+            //      <Import Project="$(MSBuildExtensionsPath)\foo\extn.proj" Condition="'Exists('$(MSBuildExtensionsPath)\foo\extn.proj')'" />
+            //
+            // The value of the MSBuildExtensionsPath* property, will always be "visible" with it's default value, example, when read or
+            // referenced anywhere else. This is a very limited support, so, it doesn't come in to effect if the explcit reference to
+            // the $(MSBuildExtensionsPath) property is not present in the Project attribute of the Import element. So, the following is
+            // not supported:
+            //
+            //      <PropertyGroup><ProjectPathForImport>$(MSBuildExtensionsPath)\foo\extn.proj</ProjectPathForImport></PropertyGroup>
+            //      <Import Project='$(ProjectPathForImport)' />
+            //
+
+            // Making a local copy for this instance of Evaluator, to add the the value of
+            // $(MSBuildExtensionsPath*) property that we have here, instead of keeping it
+            // in the static/shared MSBuildExtensionsPathHelper
+            var allExtensionPaths = new List<string>(fallbackExtensionPaths);
+            allExtensionPaths.Insert(0, _data.GetProperty(MSBuildExtensionsPathHelper.ReferenceKindToString(refKindInProject)).EvaluatedValue);
+
+            string extensionPropertyRefAsString = String.Format("$({0})", MSBuildExtensionsPathHelper.ReferenceKindToString(refKindInProject));
```

Comment:
Move this to the future struct for refKind. Something like refKindInProject.ToMSBuildPropertyName()